iotdb/src/main/java/org/apache/iotdb/db/utils/datastructure/ListPublicBAOS.java

```diff
@@ -0,0 +1,193 @@
+ * Referring to {@linkplain TVList} and {@linkplain org.apache.iotdb.db.rescon.PrimitiveArrayPool PrimitiveArrayPool}.
+ * <p>
+ * So far, this class is only used in {@linkplain org.apache.iotdb.db.engine.memtable.MemTableFlushTask
+ * MemTableFlushTask}.
+ *
+ * @author Pengze Lv, kangrong
+ */
+public class ListPublicBAOS extends PublicBAOS {
+
+  /**
+   * data buffer
+   */
+  private List<byte[]> values;
+
+  /**
+   * The number of valid bytes in the buffer.
+   */
+  private int dataSize;
+
+  /**
+   * capacity of the buffer.
+   */
+  private int capacity;
+
+  public ListPublicBAOS() {
+    super();
+
+    values = new ArrayList<>();
+    dataSize = 0;
+    capacity = 0;
+  }
+
+  public ListPublicBAOS(int size) {
+    super();
+    if (size < 0) {
+      throw new IllegalArgumentException("Negative initial size: "
+              + size);
+    }
+
+    values = ListPublicBAOSPool.getInstance().getByteLists(size);
+    dataSize = 0;
+    capacity = size;
+  }
+
+  @Override
+  public synchronized void write(byte[] b, int off, int len) {
+    if (len == 0) return;
+    if ((off < 0) || (off > b.length) || (len < 0) ||
+            ((off + len) - b.length > 0)) {
+      throw new IndexOutOfBoundsException();
+    }
+    ensureCapacity(dataSize + len);
+
+    int arrayIndex = dataSize / ARRAY_SIZE;
+    int elementIndex = dataSize % ARRAY_SIZE;
+    byte[] insertedArray = values.get(arrayIndex);
+    for (int i = off; i < off + len; i++) {
+      if (elementIndex == ARRAY_SIZE) {
+        elementIndex = 0;
+        arrayIndex++;
+        insertedArray = values.get(arrayIndex);
+      }
+      insertedArray[elementIndex] = b[i];
+      elementIndex++;
+    }
+    dataSize += len;
+  }
+
+  @Override
+  public synchronized void write(int b) {
+    ensureCapacity(dataSize + 1);
+    int arrayIndex = dataSize / ARRAY_SIZE;
+    int elementIndex = dataSize % ARRAY_SIZE;
+    values.get(arrayIndex)[elementIndex] = (byte) b;
+    dataSize++;
+  }
+
+  /**
+   * Increases the capacity if necessary to ensure that it can hold
+   * at least the number of elements specified by the minimum
+   * capacity argument.
+   */
+  private void ensureCapacity(int minCapacity) {
+    while (minCapacity - capacity > 0) {
+      values.add((byte[]) ListPublicBAOSPool.getInstance().getPrimitiveByteList());
+      capacity += ARRAY_SIZE;
+    }
+  }
+
+  public byte[] getBuf() {
+    // TODO
+    throw new UnsupportedOperationException();
+  }
+
+  public synchronized void reset() {
+    dataSize = 0;
+    if (values != null) {
+      for (byte[] dataArray : values) {
+        ListPublicBAOSPool.getInstance().release(dataArray);
+      }
+      values.clear();
+    }
+  }
+
+  public synchronized int size() {
+    return dataSize;
+  }
+
+  public void close() {
+    reset();
+  }
+
+  /**
+   * We are not sure whether following functions will be invoked in PublicBOAS. We'd defer to
+   * implement them, but throw exception to avoid unexpected calling.
+   */
```

Comment:
is this comment right?

Reply:
this comment if forgotten to delete

Reply:
> is this comment right?

this comment is forgotten to delete

Reply:
> this comment if forgotten to delete

Some functions like `toString` are not implemented yet, we could move this comment below.